Documenting the extension model of the Delivery API (#5241)
umbraco/UmbracoDocs: 12/umbraco-cms/reference/content-delivery-api/extension-points/custom-property-editors-support.md

---
description: >-
  Discover how to customize the Content Delivery API's response for your custom
  property editors.
---

# Custom property editors support

Out of the box, the Delivery API supports custom property editors, ensuring they are rendered alongside the built-in ones in Umbraco. However, if the output generated by your property editor isn't optimal for a headless context, you have the ability to customize the API response. This customization won't impact the Razor rendering, allowing you to tailor the Content Delivery API response according to your specific requirements.

> In this article, we'll look into how you can work with the IDeliveryApiPropertyValueConverter interface and implement custom [output expansion](../#output-expansion) for your custom property editors.

Reviewer: ```suggestion
In this article, we'll look into how you can work with the `IDeliveryApiPropertyValueConverter` interface and implement custom [output expansion](../#output-expansion) for your custom property editors.
```

## Prerequisite

> For the purpose of this example, it is assumed that you have created a custom property editor that stores the key of a content item. We will not dive into the details of creating a custom property editor and the necessary implementation steps within Umbraco for this article. If you need guidance on that, please refer to the [Creating a Property Editor](http://localhost:5000/s/OdQETpqkO0Kcv8KMquKL/tutorials/creating-a-property-editor) and [Property Value Converters](http://localhost:5000/s/OdQETpqkO0Kcv8KMquKL/extending/property-editors/property-value-converters) articles.

Reviewer: ```suggestion
The examples in this article revolve around the fictional property editor My.Custom.Picker. This property editor stores the key of a single content item, and is backed by a property value converter.

We will not dive into the details of creating a custom property editor for Umbraco in this article. If you need guidance on that, please refer to the [Creating a Property Editor](http://localhost:5000/s/OdQETpqkO0Kcv8KMquKL/tutorials/creating-a-property-editor) and [Property Value Converters](http://localhost:5000/s/OdQETpqkO0Kcv8KMquKL/extending/property-editors/property-value-converters) articles.
```

## Implementation

To implement the custom data values for your output in the Delivery API, we need to opt-in by implementing the `IDeliveryApiPropertyValueConverter` interface.

> The code example below showcases the implementation of a property editor called _"My.Custom.Picker"_. Our focus will be on the methods provided by the `IDeliveryApiPropertyValueConverter`, as they are responsible for customizing the Delivery API response. Towards the end of the file, you will find the response models that we will be using. The `IsConverter()` and `GetPropertyValueType()` methods are inherited from the `PropertyValueConverterBase` class, which is covered in the [Property Value Converters](http://localhost:5000/s/OdQETpqkO0Kcv8KMquKL/extending/property-editors/property-value-converters) article.

Author: There are some weird links to the https://docs.umbraco.com/umbraco-cms/tutorials/creating-a-property-editor https://docs.umbraco.com/umbraco-cms/extending/property-editors/property-value-converters/ articles. But I couldn't find them in the 12.latest space, so I had to link them through v11 folder which resulted in this. Some help from the Docs team is appreciated here!

P.S. I have no idea why it says that a sentence in ln.21 is longer than 24 words. I went though each of them with a word count

        bool expanding)
    {
        if (inter is null)
        {
            return null;
        }

        return BuildDeliveryApiCustomPicker(inter, expanding);
    }

    private DeliveryApiCustomPicker? BuildDeliveryApiCustomPicker(object inter, bool expanding)
    {
        if (!_publishedSnapshotAccessor.TryGetPublishedSnapshot(out IPublishedSnapshot? publishedSnapshot) ||
            publishedSnapshot?.Content is null)
        {
            return null;
        }

        if (!Guid.TryParse(inter as string, out Guid id))
        {
            return null;
        }

        return new DeliveryApiCustomPicker { Id = id };
    }
}

public class DeliveryApiCustomPicker
{
    public Guid Id { get; set; }

    public DeliveryApiItemDetails? ItemDetails { get; set; }
}

public class DeliveryApiItemDetails
{
    public string? Name { get; set; }

    public IApiContentRoute? Route { get; set; }
}
```
{% endcode %}

The Implementation of the `IDeliveryApiPropertyValueConverter` interface can be found in the following methods:

* `GetDeliveryApiPropertyCacheLevel()`: This method specifies the cache level used for our property representation in the Delivery API response.

> * `GetDeliveryApiPropertyValueType()`: This method defines the value type of the custom property for the Delivery API response. The value type for the API may differ from the type used for Razor rendering _(as specified by `GetPropertyValueType()`)_.

Reviewer: ```suggestion
* `GetDeliveryApiPropertyValueType()`: This method defines the value type of the custom property output for the Delivery API response.
```

Everything can differ from the Razor rendering - that's the whole point. Is there any reason to single out the value type here?

Author: You are absolutely right! There wasn't really any particular reason for it, just thought it is an easy-to-relate example 😁

>   * In the given example, having only the `Guid` as the value is sufficient for the Razor rendering. Since Razor provides full access to the published content within its rendering context, which is not the case with the Content Delivery API.

Reviewer: ```suggestion
```

Suggest to move this below (and rephrase) - see comment below.

* `ConvertIntermediateToDeliveryApiObject()`: This method converts the value from the property editor to the desired custom object in a headless context.

> 

Reviewer: ```suggestion

In the given example, the content key (`Guid` value) is used when rendering with Razor. This is sufficient, because Razor provides full access to the published content within the rendering context. However, in a headless context we do not have the same access. To prevent subsequent round-trips to the server, we create a richer output model specifically for the Delivery API.

```

Author: Makes a good of sense!

---
description: >-
  Discover how to customize the Content Delivery API's response for your custom
  property editors.
---

# Custom property editors support

Out of the box, the Delivery API supports custom property editors, ensuring they are rendered alongside the built-in ones in Umbraco. However, if the output generated by your property editor isn't optimal for a headless context, you have the ability to customize the API response. This customization won't impact the Razor rendering, allowing you to tailor the Content Delivery API response according to your specific requirements.

In this article, we'll look into how you can work with the `IDeliveryApiPropertyValueConverter` interface and implement custom [output expansion](../#output-expansion) for your custom property editors.

## Prerequisite

The examples in this article revolve around the fictional `My.Custom.Picker` property editor. This property editor stores the key of a single content item and is backed by a property value converter.

> We will not dive into the details of creating a custom property editor for Umbraco in this article. If you need guidance on that, please refer to the [Creating a Property Editor](http://localhost:5000/s/OdQETpqkO0Kcv8KMquKL/tutorials/creating-a-property-editor) and [Property Value Converters](http://localhost:5000/s/OdQETpqkO0Kcv8KMquKL/extending/property-editors/property-value-converters) articles.

Reviewer 2: ```suggestion
We will not dive into the details of creating a custom property editor for Umbraco in this article. If you need guidance on that, please refer to the [Creating a Property Editor](https://docs.umbraco.com/umbraco-cms/tutorials/creating-a-property-editor) and [Property Value Converters](https://docs.umbraco.com/umbraco-cms/extending/property-editors/property-value-converters) articles.
```

## Implementation

To customize the output of a property value editor in the Delivery API, we need to opt-in by implementing the `IDeliveryApiPropertyValueConverter` interface.

The code example below showcases the implementation of this interface in the property value converter for `My.Custom.Picker`. Our focus will be on the methods provided by the `IDeliveryApiPropertyValueConverter`, as they are responsible for customizing the Delivery API response.

Towards the end of the example, you will find the response models that we will be using.

> The `IsConverter()` and `GetPropertyValueType()` methods are inherited from the `PropertyValueConverterBase` class, which is covered in the [Property Value Converters](http://localhost:5000/s/OdQETpqkO0Kcv8KMquKL/extending/property-editors/property-value-converters) article.

Reviewer 2: ```suggestion
The `IsConverter()` and `GetPropertyValueType()` methods are inherited from the `PropertyValueConverterBase` class, which is covered in the [Property Value Converters](https://docs.umbraco.com/umbraco-cms/extending/property-editors/property-value-converters) article.
```
